### PR TITLE
Throw error to warn of mistaken loading of prod + dev React bundles

### LINKED ISSUE #9589: Ensure flat bundles don't duplicate code with weak minifiers
If the minifier doesn't eliminate constant expressions (like Uglify does), with the way our flat bundles currently work you would ship two bundles.

We should warn prominently in this case, as you would effectively ship a double bundle to production.

### PR DESCRIPTION
**Pull request by flarnie**
**what is the change?:**
Adds an error indicating a gotcha where users end up loading both the 'prod' and 'dev' bundles in production.

Credit to @gaearon for coming up with a clever way to check for this. :)

I mainly just did the manual testing and fixed a couple of typos in his
idea.

**I'd like to do a follow-up PR to add a link and a page explaining this
issue more and suggesting how to fix it. Would folks support that?**

**why make this change?:**

We want to warn for an unfortunate gotcha for
the following situation -
1. Wanting to shrink their JS bundle, an engineer runs it through
   'uglifyjs' or some other minifier. They assume this will also do
   dead-code-elimination, but the 'prod' and 'dev' checks are not
   envified yet and dev-only code does not get removed.
2. Then they run it through browserify's 'envify' or some other tool to
   change all calls to 'process.env.NODE_ENV' to "production". This
   makes their code ready to ship, except the 'dev' only dead code is
   still in the bundle. Their bundle is twice as large as it needs to
   be, due to the dead code.

This was a problem with the old build system before, but with our new
build system output it's possible to detect and throw an informative
error in this case.

**test plan:**
1. run the build in react as usual; `yarn build`
2. manually run 'uglifyjs -mt' on 'build/packages/react/index.js' to
   simulate mistakenly minifying React before env variables are
   resolved, which skips dead code elimination.
3. run the fixtures build - `cd fixtures/packaging && node
   ./build-all.js && serve ../..`
4. Visit just the production browserify fixture -
   http://localhost:5000/fixtures/packaging/browserify/prod/
5. You should see the error thrown indicating this problem has occurred.
screen shot 2017-08-11 at 4 07 44 pm
6. Do the above steps with NO uglifyjs step, and verify that no error is
   thrown. When there is no minification applied, we don't assume that
   this mix-up has occurred.
screen shot 2017-08-11 at 4 10 56 pm

**issue:**
fixes #9589

### CODE REVIEW

**Comment by acdlite** [comment]
Should we create a specific fixture for this case, where `yarn build` minifies without DCE as described in step 2? Would make it easier to test in the future.

**Comment by flarnie** [comment]
> Should we create a specific fixture for this case, where yarn build minifies without DCE as described in step 2? Would make it easier to test in the future.

I like that, and we could even point to that if we add docs about this issue in the future. Would rather merge this and add that as a follow-up, so that it unblocks the RC. But I'll work on adding a fixture over the weekend if I can.

File: packages/react/index.js [hunk 1]
```diff
@@ -1,6 +1,36 @@
...
+        // We know for a fact the above line exists.
+        // Therefore the browser gave us invalid source.
+        return;
+      }
+      if (source.indexOf('unreachable') !== -1) {
+        // Dead code elimination would have stripped that branch
+        // because it is impossible to reach in production.
+        setTimeout(function() {
+          // Ensure it gets reported to production logging
+          throw new Error('React is running in production mode, but dead code elimination has not been applied.');
```
Comment (gaearon) [inline packages/react/index.js]: This should probably include "Read how to correctly configure React for production: " and an fame link to "Use the Production Build" in "Optimizing Performance" doc.

File: packages/react/index.js [hunk 2]
```diff
@@ -1,6 +1,36 @@
 'use strict';
 
+function testMinification() {
+  if (testMinification.name === 'testMinification') {
+    // We are not minified.
+    // This might be a Node environment where DCE is not expected anyway.
+    return;
+  }
+  if (process.env.NODE_ENV === 'production') {
```
Comment (gaearon) [inline packages/react/index.js]: This outer wrapper is probably unnecessary since we are already calling it from the same condition.

**Comment by flarnie** [comment]
Re-tested manually:

With the uglify step run before envify:

screen shot 2017-08-12 at 1 50 19 pm

With no uglify step:

screen shot 2017-08-12 at 1 56 28 pm

I'd like to land this before Monday AM if possible, and will work on adding a fixture to reproduce/test it after landing.

File: packages/react/index.js [hunk 3]
```diff
@@ -1,6 +1,38 @@
...
+    throw new Error('This is unreachable');
+  }
+  try {
+    const source = testMinification.toString();
+    if (source.indexOf('toString') === -1) {
+      // We know for a fact the above line exists.
+      // Therefore the browser gave us invalid source.
+      return;
+    }
+    if (source.indexOf('unreachable') !== -1) {
```
Comment (erikdesjardins) [inline packages/react/index.js]: This is always true because it will find its own search string, i.e.
```js
(function foo() { return foo.toString().indexOf('any arbitrary string') !== -1; })();
```
so this error will be thrown in correctly minified code.

Extracting it to a variable might work:
```js
var searchString = 'any arbitrary string';
(function foo() { return foo.toString().indexOf(searchString) !== -1; })();
```
but Babili [will inline it][inlining], recreating the original problem.

`'UNREACHABLE'.toLowerCase()` appears to work for now, but probably not in the future, as Babili [currently evaluates](https://github.com/babel/minify/tree/c40dab9760fc57cd876f0d5a1e1cba0b2ac5e9ff/packages/babel-plugin-minify-constant-folding) some other `String.prototype` methods.

[inlining]: https://babeljs.io/repl/#?babili=true&evaluate=true&lineWrap=true&presets=&targets=&browsers=&builtIns=false&debug=false&experimental=true&loose=false&spec=false&code_lz=BQMwrgdgxgLglgewsAlAAgN4FgBQuBuAhgE5oDOApiVABYDKMxcEA5mgLxoDkkxVthAEYAbClwDcuXOGjwkaGBTIwAsszgg4UQnOTpsONGg1pgi5WogatOxBAB0EQgFsKHdpy7nV6zdt1c-rhGRgD0oWgA6m4kbhAIMGjOvnAUACb2wSHhaAAqNHBkSXAsNImCMWgAcghpbhQQ-HDESK4QiQDuNBR8aAAiAMIAosZF8YkUAB4ADhSw6WiEEACeHYTLmYYhfDBgxBCSWwC-WSbAXNMtaWCwdlzunnX4FMII020wgZhZYRHRaFNZrAFAUiq4YDRamgkMJlgoEGgKmhtMJRGljBA0JdajddJsQiCWh00BAKMShsQWsRzvlCqM0Lx-DQhKJAocjCctow4QYCVAkMpyAg9lA3JxvJZrP47PYYAgGExWKh2SEzmRhcRRfZmHVJgB5EDnOUK5gsL4eTgAWgAjEEtgScv8ANbxYkgBCkQhoECEYEQmKCBDPNDCZj1SaFGBkfEEtA5fI9Cju3r-xFEyikFiEYNgIrMIih9HqkUUGMEnZ7A4_NCcglqjVanVTA3ASjUeiMU3oACEHjQNrtsbjET6VHR_LqANDyScujQHWFwnRzODyiY01m6IhOjTS1o1d-iLmhFzbjgiTpcGc0wQZDIcBEbjlaD4vpoGKxV1xMoP5AoMFyK8KGFGBQEgW4kFQb57VjHIhggMg9jPRIWH_Io-BvYhFC3BFsWuCDMVeFgWFNX8jAhIkSTJNAKSpc4ACV-AvdDICsVgPzw795GcWoKAAGkRMBEjqQhx147gyKHKSCQAam4F4r2YWx5GZMYEiPBpFg3UN0nsNBGNEtBIWJZ9-UpOYYFhCSYOk2yQjkrh-QgTQWCQ_SmO9D1PxxAiAC5rLswL7O4MoYGmMhfPCEBBD2YR7H5ZxQlfWBLVmYgQEtU9LX9VKvwIy0YrgJc2V_I4UBVDksiOZEdFoUwKH0TlOVwCUUmlSDyqkHAyuVXAgA

File: packages/react/index.js [hunk 4]
```diff
@@ -1,6 +1,38 @@
...
+    // This might be a Node environment where DCE is not expected anyway.
+    return;
+  }
+  if (process.env.NODE_ENV === 'development') {
+    // We expect this method only to be called in production.
+    throw new Error('This is unreachable');
+  }
+  try {
+    const source = testMinification.toString();
+    if (source.indexOf('toString') === -1) {
```
Comment (erikdesjardins) [inline packages/react/index.js]: Similar thing here, this works, but is kinda misleading.

`const source = testMinification.toString();` could be replaced with `const source = String(testMinification);` and this check would still work.

File: packages/react/index.js (hunk 3 above)
Comment (flarnie) [inline packages/react/index.js]: ! Good catch! I'll update it. Maybe we can have it look for two occurrences of the search string.

**Comment by flarnie** [comment]
I'm thinking of pulling `testMinification` out into a separate file and writing a unit test for it, which would allow us to keep an automated test of how this works when actually run through `uglifyjs` and `babili`.  

That seems like it would add a bit of complexity to the build process - right now we don't expect this file to have access to anything except the two react bundles. I don't know if we want to maintain another file alongside the 'index.js' that gets copied over in the same way. I think it would be worthwhile to have a unit test for this logic though.

For now I'm just writing/running the unit test locally to verify this change, and if this makes sense I can figure out how to make this change to the build process.

**Comment by acdlite** [comment]
+1 to making this a unit test

**Comment by acdlite** [comment]
But I actually don't mind if it runs with Jest. If it's a separate command, that's fine, as long as it's automated and is part of CI.

**Comment by flarnie** [comment]
Pushed the WIP test - I'm not sure what config I'm missing to get `babili` to minify the method, so putting this up in case someone knows. Based on the output [here](http://babeljs.io/repl/#?babili=true&evaluate=true&lineWrap=true&presets=react%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3&targets=&browsers=&builtIns=false&debug=false&code_lz=G4QwTgBADmD2DGBTAzsiBeCBvRA7YAXNgHIDyAIgKID6lxAakQOQywAmArvAC4CWsuJgF8hAbgBQ4gGYdcPfrgjcU3ALK9cvKb3gg-AgKrJEbcgGFKACgCU2cRAjwByWABtEAOlewA5paYARuAAfgCE4RGRUdEx4UzWEg5aEJasSKgeeMAeZFS0DBjomCxwnPIC8XYODk64yNwQLhxgSBhKKuqa2rr6uEYm5pQe3LAAytxgGn4J9tW19RDeuD704LwgAe7EIAC2iG1NLYiJ1cmWSytrG1u7-wBkd7PV1YdIHjt68AAWlgD0F6tJtdENs9r8fNYvHgfNwvoVMABmWxYJ7VX6_CAAdX24H2uFgDR2Gi0vBMHlRDnREAAKl9eGgiT4vg0AjiIMR2Pssrw4Lg9rgGgB3L6IMD7QYQekQfENRAADygiB4JggIFwAE9BSB1eTng4xdxmrgTg4hKj5m5PN4_CUEChkJl8DkKDQ6PRJWgmAAaaBwdIOrLOvJumbPM5pe2O7K5V0FIrFNiIYCIbxQfncSoovUQKnYiDyxU8JR0hmIWHsCACVzqpSwCCsxwgVzuNiSxSsMq9XV62FwQXSxD9yhgOBgfy0qVS2RikDfYHxE0QM3PCY1rN6i3uLy-fyrqYeA8LimSqQpV6eD7cb5_EbjSbLcGQ9zLWEQUJFCAAJmRx5qzkt242lqGahtmOYYnmTZauqaDeLAADWEBSLAkDcIKdaXt8KAEL-4E0iKqo8BwTaVvA8DNHgrRqq2sJ4MW-y6M2tYQEwmFfEwuHopxEH7PB-L9shkAgEhs4NLRqoBLAyaLBoXJyvS3BoC49Fio0XywBwritg2n7cRAbEoN22ZUrSoqIIJ-ziQEfbGJAPggNJHBoBooCuLwrbnkZeoGkai6mseZyeWxfzTogs5fMCj5Qi-cLvpg35VGBuY4q40GwbACFIShSjofpnwisgOFgXhpmEYaJEIORYpyDiuA0SKijiYxrjMax-UcXp5Bha2TiJvmblErgegKBA6Gaa2EXSfUkxQIq9V6PWYBqt8emsroTn7LwDSTjsUCwKgvCbJZdYzt8ba-uwXBdrhm5WjuTAbBpYl1sY3DUrwexPUexWve9n0cNwlgyHIvQ2IlxW3YB_jIbAYSxPDCNRN9xV4ZQdQUZKDQ-GWaBintYDKDRdYdldI3Wj4Uy4Q4vawP2uCDhAw6jv4ABKYVFlKYCyJoyznST5SKDsnI-gEAMQImIA9ZyLFUyjxUANQsSmH0aMNAgQBFaAyvWiB0SAs1uWSEBs5LGu08xTgjkq3DVjLcv23qitMLU2g-BjJtFoJF2dgozCyw7zxO8y3BQIV6JSKLYCuB4Tg7L8p3cAAtIqYBSInG2J7RyelKTAiJ6LvBacj2ZCKBzzLtUQiNlecKWIgyIV2aZrKPUnQkj0Cj9KYFg2KIQA) I think this will work though.

**Comment by flarnie** [comment]
I could use some context on the build process and where to start when making a change to it in order to add a new file to the `build/packages/react/cjs`.

**Comment by flarnie** [comment]
I'm also not seeing the same test failures locally that CI is showing. Will look into this more on Monday.

screen shot 2017-08-13 at 6 16 36 pm

**Comment by gaearon** [comment]
The last commit doesn't feel right to me. IMO we don't want to add more special cases like this that make our bundling more complicated. Especially since this particular file is not part of bundling process.

Wouldn't it be enough to add this file to `packages/react/cjs` (and whitelist it in `files` in `package.json`)?

Although at that point it seems like a lot of complexity just to test something. IMO this is one of those cases where we need to get the semantics right, document it, and we'll likely not need to touch it again. I don't think it's worth complicating the build process or test infra to verify this function works as expected.

**Comment by flarnie** [comment]
> Although at that point it seems like a lot of complexity just to test something.

Yea, that was what I was thinking when I commented earlier asking whether this was a good direction to go in - "That seems like it would add a bit of complexity to the build process - right now we don't expect this file to have access to anything except the two react bundles."

Writing the test in the first place helped catch another bug and validate that the current version works, so I think it's worthwhile whether we decide to keep it or not.

> Wouldn't it be enough to add this file to packages/react/cjs (and whitelist it in files in package.json)?

Let me see if that works, and if not then let's consider landing this without the unit test and revisiting whether it's worth adding extra complexity.

**Comment by gaearon** [comment]
Either of two approaches would be easier to understand to me:

1. Leave it inlined in `index.js`. Sure, it needs some testing, but I doubt we'll be changing that logic since it's not *really* dependent on underlying tools (such as Uglify) and tests basic dead code elimination capabilities. No changes in Uglify or bundlers can break it unless they also break DCE (which is exactly what we want to catch).

2. If you feel strongly about unit testing it, we could put it in `packages/react/cjs`, so that it ends up in the final build in that folder. We will also need to add this file to whitelist in `package.json`. Finally, we should be able to import it from Jest with a path like `../../packages/react/cjs/...`. Although this feels hacky it would work and avoid complicating build infra for the sake of testing a single file.

**Comment by gaearon** [comment]
>Writing the test in the first place helped catch another bug and validate that the current version works, so I think was worthwhile to write it, whether we decide to keep it or not.

Totally agree! I just don't see us changing this code often once we've ironed out the first mistakes. We have similar (well, not quite as sophisticated) code for detecting some other minification mistake in Stack, and we also don't have unit tests for it. So far we've never changed it since it was written.

**Comment by flarnie** [comment]
Nearly have an update ready but I'm switching to getting https://github.com/facebook/react/pull/10385 landed and running a sync. Will come right back to this.

**Comment by flarnie** [comment]
Re-did manual local testing - 

With the minification and no dead code elimination:
screen shot 2017-08-16 at 7 08 50 am

With no minification:
screen shot 2017-08-16 at 7 27 48 am

File: yarn.lock [hunk 5]
```diff
@@ -783,6 +813,14 @@ babel-preset-jest@20.1.0-delta.1:
   dependencies:
     babel-plugin-jest-hoist "20.1.0-delta.1"
 
+babel-preset-minify@0.0.0:
+  version "0.0.0"
```
Comment (gaearon) [inline yarn.lock]: Should we revert changes to Yarn lock? Seems like Uglify update could also go into a separate PR with just necessary lockfile changes.

File: packages/react/index.js [hunk 6]
```diff
@@ -1,6 +1,50 @@
 'use strict';
 
+function testMinificationUsedDCE() {
+  // use scoped variable for our initial test, in case
+  // 'top-level' mangling is not enabled.
+  const source = testMinificationUsedDCE.toString();
```
Comment (gaearon) [inline packages/react/index.js]: I wanted this line to be inside try catch in case toString throws. I think this can happen in restrictive environments although not sure where I saw that.

File: yarn.lock (hunk 5 above)
Comment (flarnie) [inline yarn.lock]: Ack - yes, both of those should be removed from this diff. Will fix that.

File: packages/react/index.js (hunk 6 above)
Comment (flarnie) [inline packages/react/index.js]: Sure thing - I forget why I moved it out, will put it back.

File: yarn.lock (hunk 5 above)
Comment (gaearon) [inline yarn.lock]: It would still be nice to update it separately though :-) maybe it has improved in the meantime.
